### PR TITLE
CA-204242: save the server status report blocks xencenter

### PR DESCRIPTION
**Pull request by liulinC**
Signed-off-by: Lin Liu <lin.liu@citrix.com>

The save/upload process/wizard block the usage of xencenter, 
this commit extract the save/upload into an async action, which would run in background.

The async contain two sub-action, namely zip and upload.
The upload action depends on the zip action, so we need to sync the actions.

### CODE REVIEW

**Comment by liulinC** [comment]
Test done:

manually verify.

**Comment by liulinC** [comment]
@kc284  your comments make sense and refined accordingly,
could you please help to review it again?

**Comment by MihaelaStoica** [comment]
Apart from my comments above, I was also wondering which actions would be best to show in the Events: the multiple action or the two sub actions? 
For example, if saving the report succeeds but the upload fails, I think it would be better if the user sees this as two separate entries. With the current implementation they will just see that the action "Saving Server Status Report" failed while uploading, without any indication that saving the report actually succeeded.

File: XenAdminTests/XenAdminTests.csproj [hunk 1]
```diff
@@ -485,7 +485,9 @@
       <Name>XenServerHealthCheck</Name>
     </ProjectReference>
   </ItemGroup>
-  <ItemGroup />
+  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (liulinC) [reply XenAdminTests/XenAdminTests.csproj]: I think this one is added by VS automatically, not sure why VS add it, 
will delete it

File: XenModel/Actions/ZipStatusReportAction.cs [hunk 2]
```diff
@@ -62,13 +62,29 @@ public class ZipStatusReportAction : AsyncAction
...
             _inputTempFolder = tempFolder;
             _destFile = destFile;
         }
 
+        /// <summary>
+        /// Construcor, set suppressHistory to ture
+        /// </summary>
+        /// <param name="tempFolder">tempFolder to store the downloaded logs</param>
+        /// <param name="destFile">The target file to store the compress result</param>
+        public ZipStatusReportAction(string tempFolder, string destFile):this(tempFolder,destFile,true)
```
Comment (MihaelaStoica) [inline XenModel/Actions/ZipStatusReportAction.cs]: There is no need for a separate constructor, we can set the default vlaue of the suppressHistory parameter to true in the previous constructor: 
```public ZipStatusReportAction(string tempFolder, string destFile, bool suppressHistory = true)```

File: XenModel/Messages.resx [hunk 3]
```diff
@@ -744,6 +744,9 @@
   <data name="ACTION_SAVE_SETTINGS" xml:space="preserve">
     <value>Save settings for '{0}'...</value>
   </data>
+  <data name="ACTION_SAVE_UPLOAD_SERVER_STATUS_REPORT" xml:space="preserve">
```
Comment (MihaelaStoica) [inline XenModel/Messages.resx]: As we are not showing the multiple action in the Events, the action description won't be visible to the user, so we don't necessarily need new resource strings for it. I think we can reuse Messages.BUGTOOL_SAVING for startDescription and Messages.COMPLETED for endDescription

File: XenModel/Actions/HealthCheck/UploadServerStatusReportAction.cs [hunk 4]
```diff
@@ -77,12 +77,16 @@ protected override void Run()
...
+            if (Cancelling || Cancelled)
+            {
+                throw new CancelledException();
+            }
+              
             else if (string.IsNullOrEmpty(uploadUuid))
             {
                 // Fail to upload the zip to CIS server.
                 log.ErrorFormat("Fail to upload the Server Status Report {0} to CIS server {1}", bundleToUpload, UPLOAD_DOMAIN_NAME);
+                Description = Messages.ACTION_UPLOAD_SERVER_STATUS_REPORT_FAILED;
```
Comment (MihaelaStoica) [inline XenModel/Actions/HealthCheck/UploadServerStatusReportAction.cs]: Setting the Description here is not needed, as throwing the exception should be enough.

File: XenModel/Actions/ZipStatusReportAction.cs [hunk 5]
```diff
@@ -197,6 +213,7 @@ protected override void Run()
                 ZipToOutputFile(_inputTempFolder);
                 PercentComplete = 100;
                 log.ErrorFormat("An exception was trapped while creating a server status report: " + exn.Message);
+                Description = Messages.STATUS_REPORT_ZIP_FAILED;
```
Comment (MihaelaStoica) [inline XenModel/Actions/ZipStatusReportAction.cs]: Setting the Description here is not needed, as throwing the exception should be enough.

File: XenModel/Actions/ZipStatusReportAction.cs [hunk 6]
```diff
@@ -62,8 +62,14 @@ public class ZipStatusReportAction : AsyncAction
 
         private long bytesToCompress = 1;
 
-        public ZipStatusReportAction(string tempFolder, string destFile)
-            : base(null, Messages.BUGTOOL_SAVING, Messages.BUGTOOL_SAVING, true)
+        /// <summary>
+        ///  Constructor to zip downloaded file action
+        /// </summary>
+        /// <param name="tempFolder">empFolder to store the downloaded logs</param>
```
Comment (kc284) [inline XenModel/Actions/ZipStatusReportAction.cs]: "t" missing from tempFolder
"compressed" result
whether "to" suppress history in the "events tab"

File: XenAdmin/Wizards/BugToolWizardFiles/BugToolWizard.cs [hunk 7]
```diff
@@ -120,33 +121,28 @@ protected override void FinishWizard()
...
-            // zip up the report files and save them to the chosen file
-            Actions.ZipStatusReportAction action =
-                new Actions.ZipStatusReportAction(bugToolPageRetrieveData.OutputFolder, bugToolPageDestination1.OutputFile);
-            using (var dialog = new ActionProgressDialog(action, ProgressBarStyle.Blocks))
+            AsyncAction action;
+            if (bugToolPageDestination1.Upload)
             {
-                dialog.ShowCancel = true;
-                dialog.ShowDialog();
+                // The MultipleAction only used as a wrapper, we will suppress its history and expose the sub-actions in the history
```
Comment (kc284) [inline XenAdmin/Wizards/BugToolWizardFiles/BugToolWizard.cs]: The MultipleAction "is" only...

File: XenModel/Actions/ZipStatusReportAction.cs (hunk 6 above)
Comment (liulinC) [reply XenModel/Actions/ZipStatusReportAction.cs]: updated

File: XenAdmin/Wizards/BugToolWizardFiles/BugToolWizard.cs (hunk 7 above)
Comment (liulinC) [reply XenAdmin/Wizards/BugToolWizardFiles/BugToolWizard.cs]: updated

**Pull request merged**